topology/intel/ssp-nhlt.c

```diff
@@ -0,0 +1,860 @@
+		 sscr2, sspsp2, sscr3, ssioc);
+	dai_info(dai, "ssp_set_config(), ssrsa = 0x%08x, sstsa = 0x%08x",
+		 ssrsa, sstsa);
+
+	ssp->state[DAI_DIR_PLAYBACK] = COMP_STATE_PREPARE;
+	ssp->state[DAI_DIR_CAPTURE] = COMP_STATE_PREPARE;
+
+out:
+	platform_shared_commit(ssp, sizeof(*ssp));
+
+	spin_unlock(&dai->lock);
+
+	return ret;
+}
+
+/*
+ * Portion of the SSP configuration should be applied just before the
+ * SSP dai is activated, for either power saving or params runtime
+ * configurable flexibility.
+ */
+static int ssp_pre_start(struct dai *dai)
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+	struct sof_ipc_dai_config *config = &ssp->config;
+	uint32_t sscr0;
+	uint32_t mdiv;
+	bool need_ecs = false;
+
+	int ret = 0;
+
+	dai_info(dai, "ssp_pre_start()");
+
+	/* SSP active means bclk already configured. */
+	if (ssp->state[SOF_IPC_STREAM_PLAYBACK] == COMP_STATE_ACTIVE ||
+	    ssp->state[SOF_IPC_STREAM_CAPTURE] == COMP_STATE_ACTIVE)
+		return 0;
+
+	/* MCLK config */
+	ret = mn_set_mclk(config->ssp.mclk_id, config->ssp.mclk_rate);
+	if (ret < 0) {
+		dai_err(dai, "invalid mclk_rate = %d for mclk_id = %d",
+			config->ssp.mclk_rate, config->ssp.mclk_id);
+		goto out;
+	}
+
+	sscr0 = ssp_read(dai, SSCR0);
+
+#if CONFIG_INTEL_MN
+	/* BCLK config */
+	ret = mn_set_bclk(config->dai_index, config->ssp.bclk_rate,
+			  &mdiv, &need_ecs);
+	if (ret < 0) {
+		dai_err(dai, "invalid bclk_rate = %d for dai_index = %d",
+			config->ssp.bclk_rate, config->dai_index);
+		goto out;
+	}
+#else
+	if (ssp_freq[SSP_DEFAULT_IDX].freq % config->ssp.bclk_rate != 0) {
+		dai_err(dai, "invalid bclk_rate = %d for dai_index = %d",
+			config->ssp.bclk_rate, config->dai_index);
+		goto out;
+	}
+
+	mdiv = ssp_freq[SSP_DEFAULT_IDX].freq / config->ssp.bclk_rate;
+#endif
+
+	if (need_ecs)
+		sscr0 |= SSCR0_ECS;
+
+	/* clock divisor is SCR + 1 */
+	mdiv -= 1;
+
+	/* divisor must be within SCR range */
+	if (mdiv > (SSCR0_SCR_MASK >> 8)) {
+		dai_err(dai, "ssp_pre_start(): divisor %d is not within SCR range",
+			mdiv);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* set the SCR divisor */
+	sscr0 &= ~SSCR0_SCR_MASK;
+	sscr0 |= SSCR0_SCR(mdiv);
+
+	dai_info(dai, "ssp_set_config(), sscr0 = 0x%08x", sscr0);
+out:
+	platform_shared_commit(ssp, sizeof(*ssp));
+
+	return ret;
+}
+
+int set_ssp_nhlt(struct tplg_pre_processor *tplg_pp,
+		  snd_config_t *cfg, snd_config_t *parent)
+{
+	snd_config_t *items;
+	int ret;
+	snd_config_iterator_t i, next;
+	snd_config_t *n;
+	const char *id;
+
+	tplg_pp_debug("set_ssp_nhlt");
+
+	set_dmic_data(tplg_pp, cfg);
+
+	ret = snd_config_search(cfg, "Object.Base.pdm_config", &items);
+	if (ret < 0)
+		return 0;
+
+	snd_config_for_each(i, next, items) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		tplg_pp_debug("set_pdm_data for  id: '%s' ...", id);
+
+		set_pdm_data(tplg_pp, n);
+	}
+
+	dmic_build_nhlt();
+
+	return 0;
+}
```

Comment:
I think you should really split the code in two
a) new NHLT parts with the license of your choice (no objection to LGPL on my side)
b) helpers ripped from the SOF tree without a license change. that way you can add fixes later (both ways) on without having to worry about license, etc.